src/basic/process-util.c

```diff
@@ -1801,36 +1810,33 @@ int safe_fork_full(
         if (FLAGS_SET(flags, FORK_FREEZE))
                 freeze();
 
-        if (ret_pid)
-                *ret_pid = getpid_cached();
+        if (ret_pid) {
+                r = pidref_set_self(ret_pid);
+                if (r < 0) {
+                        log_full_errno(prio, r, "Failed to acquire PID reference on ourselves: %m");
+                        _exit(EXIT_FAILURE);
+                }
+        }
 
         return 0;
 }
 
-int pidref_safe_fork_full(
+int safe_fork_full(
                 const char *name,
                 const int stdio_fds[3],
                 int except_fds[],
                 size_t n_except_fds,
                 ForkFlags flags,
-                PidRef *ret_pid) {
+                pid_t *ret_pid) {
 
-        pid_t pid;
-        int r, q;
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        int r;
 
-        r = safe_fork_full(name, stdio_fds, except_fds, n_except_fds, flags, &pid);
+        r = pidref_safe_fork_full(name, stdio_fds, except_fds, n_except_fds, flags, ret_pid ? &pidref : NULL);
```

Comment:
kinda weird allocating the pidfd here if we are not going to use that, in particular in the child, where we usuall won't need it.

maybe add a new flag that suppressed pidfd allocation unless we don't have it anyway and set that here?